### DISCUSSION TITLE
Add electron includes

### DISCUSSION THREAD

**Discussion by arthuro555** [4ian/GDevelop#2600] [Enhancement ideas (and their technical discussions)]
We should probably add some sort of system to add includes to the electron main.js, as some possible extensions could only be made that way (for example, discord rich presence only works on the main process, not on the renderer process, or an extension to get command line arguments would require accessing the node global `process`, only accessible in main, etc). This could also allow the removal of the unsafe and slow `require("electron").remote` in favor of ipc events in some extensions.
How should we do this? Adding a special `addElectronInclude()` method to a gdjs `PlatformExtension` subclass would be maybe a way.

**Comment by arthuro555** [comment]
Bump, I still think that this would be very useful. Additionally, I think it would be nice to add more "include types", maybe with an optional parameter on `addInclude`. Some include types on the top of my mind:
- "node": Gets `require`'d in the node process if there is one
- "preload": Gets preloaded via [electron preload](https://www.electronjs.org/docs/latest/tutorial/process-model/#preload-scripts)
- "service": Gets imported from a service worker via `importScripts` (not possible for now as games do not have one yet)
- "manual": Gets the file exported but does not import it from anywhere. Useful for example for workers.


This would allow to 
1. Use electron the way the electron team is pushing everyone towards
2. Allowing to use stuff like Firebase Messenging or advanced cache controls
3. Workers for long background tasks

**Reply by arthuro555** [reply]
Bump, I would like to work on this and am awaiting feedback before starting to attempt any work on this

**Reply by 4ian** [reply]
I'm afraid this is quite complicated and would not work with previews - which is a big problem if we start to have one, then two, then 10 extensions that can't be previewed. Can we list use cases of extensions that would be using this? 
I could get convinced, but only if:
- the design is *super* simple. There is not a single line of code more than needed for the extensions that requires this.
- there are at least 1 or 2 extensions needing this. (i.e: we start from the need for a solution)
- previews can work with the system :) 

It might be tricky, but we would need this level of quality I think.
